server: TestCheckRestartSafe_AllowMinimumQuorum_Pass failed

server.TestCheckRestartSafe_AllowMinimumQuorum_Pass [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20324717?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20324717?buildTab=artifacts#/) on master @ [6863798d76ed9e171c7accaceea6e632e86c84bd](https://github.com/cockroachdb/cockroach/commits/6863798d76ed9e171c7accaceea6e632e86c84bd):


```
panic({0x56e8ae0?, 0x40145bf7d0?})
	GOROOT/src/runtime/panic.go:792 +0x124
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.(*Manager).SetDraining.func1(0x40047b1630, {0x6a71fb8?, 0x4012846730?}, 0x1)
	pkg/sql/catalog/lease/lease.go:1619 +0x1ac
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.(*Manager).SetDraining(0x4005f3fb88, {0x6a71fb8, 0x4012846730}, 0x1, 0x4008cfb290, 0x1)
	pkg/sql/catalog/lease/lease.go:1622 +0x154
github.com/cockroachdb/cockroach/pkg/server.(*drainServer).drainClientsInternal(0x40039c9db0, {0x6a749e0?, 0x400587bb80?}, 0x4008cfb290, 0x1)
	pkg/server/drain.go:495 +0x3c4
github.com/cockroachdb/cockroach/pkg/server.(*drainServer).drainClients(...)
	pkg/server/drain.go:402
github.com/cockroachdb/cockroach/pkg/server.(*drainServer).drainInner(0x40039c9db0, {0x6a749e0, 0x400587bb80}, 0x4008cfb290, 0x0)
	pkg/server/drain.go:383 +0x140
github.com/cockroachdb/cockroach/pkg/server.(*drainServer).runDrain(0x40039c9db0, {0x6a749e0, 0x400587bb80}, 0x0)
	pkg/server/drain.go:343 +0xf8
github.com/cockroachdb/cockroach/pkg/server.(*drainServer).handleDrain(0x40039c9db0, {0x6a749e0, 0x400587bb80}, 0x4004d0f960, {0xffff34690778, 0x401483c460})
	pkg/server/drain.go:193 +0xd4
github.com/cockroachdb/cockroach/pkg/server.(*adminServer).drain(0x4006673040, 0x4004d0f960, {0xffff34690778, 0x401483c460})
	pkg/server/drain.go:120 +0x118
github.com/cockroachdb/cockroach/pkg/server.(*adminServer).Drain(...)
	pkg/server/drain.go:91
github.com/cockroachdb/cockroach/pkg/server/serverpb._Admin_Drain_Handler({0x5c24f80, 0x40066084f8}, {0x6aa84c0, 0x400e62a1e0})
	bazel-out/aarch64-fastbuild/bin/pkg/server/serverpb/serverpb_go_proto_/github.com/cockroachdb/cockroach/pkg/server/serverpb/admin.pb.go:6180 +0x11c
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.StreamServerInterceptor.func13({0x5c24f80, 0x40066084f8}, {0x6aa84c0, 0x400e62a1e0}, 0x4010d47500, 0x6099760)
	pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:123 +0x324
google.golang.org/grpc.getChainStreamHandler.func1({0x5c24f80?, 0x40066084f8?}, {0x6aa84c0?, 0x400e62a1e0?})
	external/org_golang_google_grpc/server.go:1504 +0x124
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func4({0x5c24f80, 0x40066084f8}, {0x6aa84c0, 0x400e62a1e0}, 0x4010d47500?, 0x4011e69e00)
	pkg/rpc/context.go:189 +0x70
google.golang.org/grpc.getChainStreamHandler.func1.getChainStreamHandler.1({0x5c24f80, 0x40066084f8}, {0x6aa84c0, 0x400e62a1e0})
	external/org_golang_google_grpc/server.go:1504 +0x90
github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.streamInterceptor({0x400298a000, {{0x1}, {0x6ac69e0, 0x4005bb60c0}}, 0x0}, {0x5c24f80, 0x40066084f8}, {0x6aa84c0, 0x400e62a1e0}, 0x4010d47500, ...)
	pkg/rpc/auth.go:193 +0x308
google.golang.org/grpc.getChainStreamHandler.func1({0x5c24f80?, 0x40066084f8?}, {0x6aa84c0?, 0x400e62a1e0?})
	external/org_golang_google_grpc/server.go:1504 +0x124
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func2.1({0x0?, 0x6a71ff0?})
	pkg/rpc/context.go:146 +0x34
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr(0x40032d0870, {0x6a71ff0, 0x4015b645a0}, {0x40102e7590, 0x26}, 0x4018d7ca98)
	pkg/util/stop/stopper.go:349 +0xec
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func2({0x5c24f80, 0x40066084f8}, {0x6aa84c0, 0x400e62a1e0}, 0x4010d47500, 0x4011e69d80)
	pkg/rpc/context.go:145 +0xb4
google.golang.org/grpc.NewServer.chainStreamServerInterceptors.chainStreamInterceptors.func2({0x5c24f80, 0x40066084f8}, {0x6aa84c0, 0x400e62a1e0}, 0x4010d47500, 0x401483c450?)
	external/org_golang_google_grpc/server.go:1495 +0x7c
google.golang.org/grpc.(*Server).processStreamingRPC(0x4004282960, {0x6ac8de0, 0x4003a5c1a0}, 0x400ea82a20, 0x4001ee9ad0, 0xa4fb700, 0x0)
	external/org_golang_google_grpc/server.go:1659 +0xf6c
google.golang.org/grpc.(*Server).handleStream(0x4004282960, {0x6ac8de0, 0x4003a5c1a0}, 0x400ea82a20, 0x0)
	external/org_golang_google_grpc/server.go:1739 +0x780
google.golang.org/grpc.(*Server).serveStreams.func1.1()
	external/org_golang_google_grpc/server.go:970 +0xac
created by google.golang.org/grpc.(*Server).serveStreams.func1 in goroutine 30300
	external/org_golang_google_grpc/server.go:981 +0x148
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/obs-prs @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCheckRestartSafe_AllowMinimumQuorum_Pass.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

server.TestCheckRestartSafe_AllowMinimumQuorum_Pass [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20333469?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20333469?buildTab=artifacts#/) on master @ [36d351d47f7f35b1c741fabc32e84e41221873b6](https://github.com/cockroachdb/cockroach/commits/36d351d47f7f35b1c741fabc32e84e41221873b6):


```
=== RUN   TestCheckRestartSafe_AllowMinimumQuorum_Pass
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/077c6015fca14174d23f0b07262a46e6/logTestCheckRestartSafe_AllowMinimumQuorum_Pass541397390
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:143: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
    api_v2_test.go:460: 
        	Error Trace:	pkg/server/api_v2_test.go:460
        	Error:      	Should be true
        	Test:       	TestCheckRestartSafe_AllowMinimumQuorum_Pass
    panic.go:636: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/077c6015fca14174d23f0b07262a46e6/logTestCheckRestartSafe_AllowMinimumQuorum_Pass541397390
--- FAIL: TestCheckRestartSafe_AllowMinimumQuorum_Pass (20.13s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCheckRestartSafe_AllowMinimumQuorum_Pass.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/server/api_v2_test.go
+++ b/pkg/server/api_v2_test.go
@@ -199,18 +199,38 @@ func TestRestartSafetyV2(t *testing.T) {
 	// Drain the node; then we'll expect success.
 	require.NoError(t, drain(ctx, ts1, t))
 
-	req, err = http.NewRequest("GET", urlStr, nil)
-	require.NoError(t, err)
-	resp, err = client.Do(req)
-	require.NoError(t, err)
-	require.NotNil(t, resp)
+	testutils.SucceedsSoon(t, func() error {
+		req, err := http.NewRequest("GET", urlStr, nil)
+		if err != nil {
+			return err
+		}
+		resp, err := client.Do(req)
+		if err != nil {
+			return err
+		}
+		if resp == nil {
+			return fmt.Errorf("response is nil")
+		}
 
-	bodyBytes, err = io.ReadAll(resp.Body)
-	require.NoError(t, err)
-	require.NoError(t, resp.Body.Close())
+		bodyBytes, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return err
+		}
+		if err := resp.Body.Close(); err != nil {
+			return err
+		}
 
-	require.Equal(t, 200, resp.StatusCode, "expected 200: %s", string(bodyBytes))
-	require.NoError(t, json.Unmarshal(bodyBytes, &response))
+		if resp.StatusCode != 200 {
+			return fmt.Errorf("expected 200 but got %d: %s", resp.StatusCode, string(bodyBytes))
+		}
+
+		var response RestartSafetyResponse
+		if err := json.Unmarshal(bodyBytes, &response); err != nil {
+			return err
+		}
+
+		return nil
+	})
 }
 
 // TestRulesV2 tests the /api/v2/rules endpoint to ensure it
@@ -373,6 +393,8 @@ func TestCheckRestartSafe_RangeStatus(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
+	skip.UnderDuress(t)
+
 	ctx := context.Background()
 	var err error
 
@@ -392,14 +414,29 @@ func TestCheckRestartSafe_RangeStatus(t *testing.T) {
 	require.True(t, vitality.GetNodeVitalityFromCache(ts0.NodeID()).IsDraining())
 	require.False(t, vitality.GetNodeVitalityFromCache(ts1nodeID).IsLive(livenesspb.Metrics))
 
-	res, err := checkRestartSafe(ctx, ts0.NodeID(), vitality, ts0.GetStores().(storeVisitor), 3, false)
-	require.NoError(t, err)
-	require.False(t, res.IsRestartSafe, "expected unsafe since a different node is down")
+	testutils.SucceedsSoon(t, func() error {
+		res, err := checkRestartSafe(ctx, ts0.NodeID(), vitality, ts0.GetStores().(storeVisitor), 3, false)
+		if err != nil {
+			return err
+		}
+		if res.IsRestartSafe {
+			return fmt.Errorf("expected unsafe since a different node is down")
+		}
 
-	require.Equal(t, int32(0), res.UnavailableRangeCount)
-	require.Equal(t, int32(0), res.RaftLeadershipOnNodeCount)
-	require.Equal(t, int32(0), res.StoreNotDrainingCount)
-	require.Greater(t, res.UnderreplicatedRangeCount, int32(0))
+		if res.UnavailableRangeCount != 0 {
+			return fmt.Errorf("expected UnavailableRangeCount=0, got %d", res.UnavailableRangeCount)
+		}
+		if res.RaftLeadershipOnNodeCount != 0 {
+			return fmt.Errorf("expected RaftLeadershipOnNodeCount=0, got %d", res.RaftLeadershipOnNodeCount)
+		}
+		if res.StoreNotDrainingCount != 0 {
+			return fmt.Errorf("expected StoreNotDrainingCount=0, got %d", res.StoreNotDrainingCount)
+		}
+		if res.UnderreplicatedRangeCount <= 0 {
+			return fmt.Errorf("expected UnderreplicatedRangeCount>0, got %d", res.UnderreplicatedRangeCount)
+		}
+		return nil
+	})
 }
 
 func updateAllReplicaCounts(
@@ -455,9 +492,16 @@ func TestCheckRestartSafe_AllowMinimumQuorum_Pass(t *testing.T) {
 	require.NoError(t, err)
 
 	vitality.Draining(ts0.NodeID(), true)
-	res, err := checkRestartSafe(ctx, ts0.NodeID(), vitality, stores, testCluster.NumServers(), false)
-	require.NoError(t, err)
-	require.True(t, res.IsRestartSafe)
+	testutils.SucceedsSoon(t, func() error {
+		res, err := checkRestartSafe(ctx, ts0.NodeID(), vitality, stores, testCluster.NumServers(), false)
+		if err != nil {
+			return err
+		}
+		if !res.IsRestartSafe {
+			return fmt.Errorf("expected node to be restart safe")
+		}
+		return nil
+	})
 
 	ts1nodeID := testCluster.Server(1).NodeID()
 	vitality.DownNode(ts1nodeID)
@@ -465,10 +509,16 @@ func TestCheckRestartSafe_AllowMinimumQuorum_Pass(t *testing.T) {
 	require.True(t, vitality.GetNodeVitalityFromCache(ts0.NodeID()).IsDraining())
 	require.False(t, vitality.GetNodeVitalityFromCache(ts1nodeID).IsLive(livenesspb.Metrics))
 
-	res, err = checkRestartSafe(ctx, ts0.NodeID(), vitality, stores, testCluster.NumServers(), true)
-	require.NoError(t, err)
-
-	require.True(t, res.IsRestartSafe)
+	testutils.SucceedsSoon(t, func() error {
+		res, err := checkRestartSafe(ctx, ts0.NodeID(), vitality, stores, testCluster.NumServers(), true)
+		if err != nil {
+			return err
+		}
+		if !res.IsRestartSafe {
+			return fmt.Errorf("expected node to be restart safe with minimum quorum")
+		}
+		return nil
+	})
 }
 
 // TestCheckRestartSafe_AllowMinimumQuorum_Fail verifies that with 2

